src/components/tests/SourcesTree.spec.js

```diff
@@ -6,32 +6,617 @@ import React from "react";
+    it("renders empty custom root source list", async () => {
+      expect(emptyComponent).toMatchSnapshot();
+    });
+  });
+
+  describe("onContextMenu of the tree", () => {
+    it("shows context menu on directory to set as root", async () => {
+      const menuOptions = [
+        {
+          accesskey: "r",
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-set-directory-root",
+          label: "Set directory root"
+        }
+      ];
+      await componentWithRoot
+        .instance()
+        .onContextMenu(mockEvent, mockDirectory);
+      expect(showMenu).toHaveBeenCalledWith(mockEvent, menuOptions);
+
+      expect(mockEvent.preventDefault).toHaveBeenCalled();
+      expect(mockEvent.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(propsWithRoot.setProjectDirectoryRoot).toHaveBeenCalled();
+      expect(propsWithRoot.clearProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(copyToTheClipboard).not.toHaveBeenCalled();
+    });
+
+    it("shows context menu on file to copy source uri", async () => {
+      const menuOptions = [
+        {
+          accesskey: "u",
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-menu-copy-source",
+          label: "Copy source URI"
+        }
+      ];
+      await componentWithRoot
+        .instance()
+        .onContextMenu(mockEvent, singleMockItem);
+      expect(showMenu).toHaveBeenCalledWith(mockEvent, menuOptions);
+
+      expect(mockEvent.preventDefault).toHaveBeenCalled();
+      expect(mockEvent.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(propsWithRoot.setProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(propsWithRoot.clearProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(copyToTheClipboard).toHaveBeenCalled();
+    });
+
+    it("shows context menu on root to remove directory root", async () => {
+      const menuOptions = [
+        {
+          click: expect.any(Function),
+          disabled: false,
+          id: "node-remove-directory-root",
+          label: "Remove directory root"
+        }
+      ];
+      await componentWithRoot
+        .instance()
+        .onContextMenu(mockEvent, mockRootDirectory);
+      expect(showMenu).toHaveBeenCalledWith(mockEvent, menuOptions);
+
+      expect(mockEvent.preventDefault).toHaveBeenCalled();
+      expect(mockEvent.stopPropagation).toHaveBeenCalled();
+
+      showMenu.mock.calls[0][1][0].click();
+      expect(propsWithRoot.setProjectDirectoryRoot).not.toHaveBeenCalled();
+      expect(propsWithRoot.clearProjectDirectoryRoot).toHaveBeenCalled();
+      expect(copyToTheClipboard).not.toHaveBeenCalled();
+    });
+  });
+
+  describe("renderItem", () => {
+    it("should show icon for webpack item", async () => {
+      const item = createMockItem("webpack://", "webpack://");
+      const node = renderItem(defaultComponent, item);
```

Comment:
I will use this as an example to help make clear what i mean.

We have a test, where we are using a variable called `defaultComponent`. in order for me, an outsider, to know what this test is doing (for example, "why is this failing?", imagine being in a frustrated state of mind), we have to scroll up to line 83. There we see that `defaultComponent` can be modified, because it is an undefined `let`, meaning that it isn't the same thing all the time! Oh no!

next, we have to go through every time `defaultComponent` is defined. it is defined at line 96. then, in every test it goes through some kind of modification. So as the person reading this code, i have to remember -- ok it is in a before each, so it gets reset, are there other beforeEachs? is there some case where this is not reset? is there more nesting that i need to keep in mind? how much of the scope do i need to know about? are there other ways this could get modified?

As you can see, the mental load is _high_. I need to remember a lot, and i need to read a lot of code to know what is happening. 

Ok, now let's define what we are working with, in the test:

```javascript
 describe("renderItem", () => {
    it("should show icon for webpack item", async () => {
      const { component } = render();
      const item = createMockItem("webpack://", "webpack://");
      const node = renderItem(component, item);
      expect(node).toMatchSnapshot();
    });
```

i don't have to go anywhere, and this only costs a single extra line! `component` can mean **only one thing** here. this is a massive improvement for readability. 

DRY code is only good if it makes sense, and I would first write code out WET before drying it up. do not dry up code if it makes it harder to read, and try to keep an outsiders (often emotional and tired) perspective in mind! you always have an audience for your code :)